compaction/compaction.cc

```diff
@@ -1622,7 +1622,7 @@ class resharding_compaction final : public compaction {
     }
 
     bool use_interposer_consumer() const override {
-        return true;
+        return false;
```

Comment:
Maybe a comment here? the relationship between compaction and use_interposer_consumer() isn't clear. It just says to interpose something, but what?

Maybe it should be renamed.

Reply:
The question is if we want resharding to avoid mutation compaction altogether (not only compacting_reader which is one way), and defer the compaction work to reshape as I proposed.

This patch plays with existing logic, to avoid compacting_reader which is compaction on producer side rather than on consumer side (see compact_for_compaction_v2). At the time, I measured compacting_reader to be 3% slower than the alternative.

So switching to alternative is already an advantage, but a better way to achieve it will require some rework on compaction::consume(). There will be possibly a need for a new virtual method that will allow resharding to customize better the creation of consumer. But note there will be some overlapping with existing make_interposer_consumer()

Reply:
> The question is if we want resharding to avoid mutation compaction altogether (not only compacting_reader which is one way), and defer the compaction work to reshape as I proposed.

IIUC, we should *only* do what reshard is supposed to do. this is what @raphaelsc proposed in the description of the issue.  as compaction is not free.

> 
> This patch plays with existing logic, to avoid compacting_reader which is compaction on producer side rather than on consumer side (see compact_for_compaction_v2). At the time, I measured compacting_reader to be 3% slower than the alternative.
> 
> So switching to alternative is already an advantage, but a better way to achieve it will require some rework on compaction::consume(). 

actually, i do have a patch locally to refactor `compaction::consume()`  . thought we'd want have a smaller patch to ease the backport, so would want to send it after this change gets merged. 

> There will be possibly a need for a new virtual method that will allow resharding to customize better the creation of consumer. But note there will be some overlapping with existing make_interposer_consumer()

probably, instead of customizing the creation of consumer via the knob methods, we should make `consume()` a virtual method, so that the compaction's derived class can override it.

Reply:
> Maybe a comment here? the relationship between compaction and use_interposer_consumer() isn't clear. 

agreed. i will accompany this change with  comment explaining the connection between compaction and interposer consumer, and the connection between compaction and consumer in general.

> It just says to interpose something, but what?

in short,  the hyper consumer which consumes mutation fragments.

IIUC, we use a hyper consumer (`compact_mutation_v2`) which is plugged into a mutation reader. the hyper consumer pipes the mutation fragments fed by the reader to its downstream consumers which do the heavy lifting. as these downstream consumers write the (compacted) fragments to new sstables, they are referenced as "writer" in the code sometimes. 

two downstream consumers are be configured:

- Consumer: a regular consumer.
- GCConsumer: for consuming tombstones

```mermaid
classDiagram
      class FlattenedConsumerV2 {
          <<interface>>
          +consume_new_of_partition()
          +consume_end_of_partition()
      }
      class flat_mutation_reader_v2_impl {
          <<interface>>
          +fill_buffer()
          +next_partition()
      }
      class compacting_reader {
          - compact_mutation_state _compactor
          - flat_mutation_reader_v2 _reader
      }
      class flat_multi_range_mutation_reader {
      }
      compacting_reader --|> flat_mutation_reader_v2_impl
      flat_multi_range_mutation_reader --|> flat_mutation_reader_v2_impl
      class flat_mutation_reader_v2 {
          +consume_in_thread(FlattenedConsumerV2)
      }
      flat_mutation_reader_v2 *-- flat_mutation_reader_v2_impl
      flat_mutation_reader_v2 ..> FlattenedConsumerV2
      class compact_mutation_v2 {
          +consume(tombstone)
          +consume(static_row)
          +consume_new_of_partition()
          +consume_end_of_partition()
      }
      compact_mutation_v2 ..|> FlattenedConsumerV2
      class CompactedFragmentsConsumerV2 {
          <<interface>>
          +consume(tombstone)
          +consume(static_row)
          +consume_end_of_partition()
      }
      class compact_mutation_state {
          +consume(tombstone,Consumer,GCConsumer)
          +consume(static_row,Consumer,GCConsumer)
          +consume_new_partition()
          +consume_end_of_partition()
      }
      Consumer ..|> CompactedFragmentsConsumerV2
      GCConsumer ..|> CompactedFragmentsConsumerV2
      compact_mutation_v2 *-- Consumer
      compact_mutation_v2 *-- GCConsumer
      compact_mutation_v2 *-- compact_mutation_state
```
> 
> Maybe it should be renamed.

Reply:
@raphaelsc hi Raphel, could you shed some light on what writer we should be using for resharding's consumer? IIUC, we need to have something like
```c++
    future<> consume() override {
        auto producer = setup_sstable_reader();
        auto consumer = [this] (flat_mutation_reader_v2 reader) mutable {
            return seastar::async([this, reader = std::move(reader)] () mutable {
                auto close_reader = deferred_close(reader);
                auto now = gc_clock::now();
                writer = /// a write which create a new sstable at seeing a new partition,
                         // quite like how compact_for_compaction_v2 works
                reader.consume_in_thread(std::move(writer));
            });
        };
        return  mutation_writer::segregate_by_shard(std::move(producer), std::move(consumer));
    }
```
for the `resharding_compaction`.

if we don't have such a writer yet, i will implement one. but it seems that you implied that we only needed to customize the way how the writer/consumer is created.